runelite-client/src/main/java/net/runelite/client/plugins/puzzlesolver/solver/pathfinding/IDAStarMM.java

```diff
@@ -0,0 +1,722 @@
+
+		//Check if already placed correct
+		if (locVal.getX() == x && locVal.getY() == y)
+		{
+			return;
+		}
+
+		//Check if simple swap is enough
+		if (locBlank.getX() == x && locBlank.getY() == y && locVal.getX() - 1 == x)
+		{
+			swap(locBlank, locVal);
+			return;
+		}
+
+		//Move left
+		while (true)
+		{
+			locVal = findPiece(valTarget);
+			locBlank = findPiece(BLANK_TILE_VALUE);
+
+			//Check if already placed correct
+			if (locVal.getX() == x && locVal.getY() == y)
+			{
+				return;
+			}
+
+			if (locVal.getX() == locBlank.getX())
+			{
+				int diff = locBlank.getY() - locVal.getY();
+				if (diff == 1)
+				{
+					//Below
+					performSwapPattern(locBlank, locVal, SHUFFLE_UP_BELOW);
+				}
+				else if (diff == -1)
+				{
+					//Above
+
+					//Don't remove correct pices from row
+					if (locVal.getX() - 1 == x)
+					{
+						//Swap blank to right and continue
+						Point loc1 = new Point(locBlank.getX() + 1, locBlank.getY());
+						Point loc2 = new Point(loc1.getX(), loc1.getY() + 1);
+
+						swap(locBlank, loc1);
+						swap(loc1, loc2);
+
+						continue;
+					}
+
+					performSwapPattern(locBlank, locVal, SHUFFLE_UP_ABOVE);
+				}
+			}
+			else if (locVal.getY() == locBlank.getY())
+			{
+				int diff = locBlank.getX() - locVal.getX();
+				if (diff == 1)
+				{
+					//Right
+
+					//Last piece
+					if (y == DIMENSION - 1)
+					{
+						performSwapPattern(locBlank, locVal, LAST_PIECE_COLUMN);
+						return;
+					}
+
+					performSwapPattern(locBlank, locVal, ROTATE_LEFT_DOWN);
+				}
+				else if (diff == -1)
+				{
+					//Left
+					swap(locBlank, locVal);
+				}
+			}
+		}
+	}
+
+	private void swap(Point p1, Point p2)
+	{
+		PuzzleState newState = currentState.swap(p1.getX(), p1.getY(), p2.getX(), p2.getY());
+
+		currentState = newState;
+		stateList.add(newState);
+	}
+
+	private Point findPiece(int val)
+	{
+		for (int x = 0; x < DIMENSION; x++)
+		{
+			for (int y = 0; y < DIMENSION; y++)
+			{
+				if (currentState.getPiece(x, y) == val)
+				{
+					return new Point(x, y);
+				}
+			}
+		}
+		// This should never happen
+		throw new IllegalStateException("Piece wasn't found!");
+	}
+
+	/**
+	 * Assumes locBlank is first point for swap and locVal is last point for swap
+	 *
+	 * swap(locBlank, loc1);
+	 * swap(loc1, loc2);
+	 * swap(loc2, locVal);
+	 */
+	private void performSwapPattern(Point locBlank, Point locVal, PuzzleSwapPattern pattern)
+	{
+		int[] offsets;
+		switch (pattern)
+		{
+			case ROTATE_LEFT_UP:
+			case ROTATE_RIGHT_UP:
+			case ROTATE_RIGHT_DOWN:
+			case ROTATE_LEFT_DOWN:
+				offsets = ROTATE_LEFT_UP.getPoints();
+				break;
+			case ROTATE_UP_LEFT:
+			case ROTATE_UP_RIGHT:
+			case ROTATE_DOWN_LEFT:
+			case ROTATE_DOWN_RIGHT:
+				offsets = ROTATE_UP_LEFT.getPoints();
+				break;
+			case LAST_PIECE_ROW:
+				offsets = pattern.getPoints();
+				break;
+			case LAST_PIECE_COLUMN:
+				offsets = pattern.getPoints();
+				break;
+			default:
+				offsets = pattern.getPoints();
```

Comment:
These last 3 cases do the same thing?

Reply:
Actually the two above can be merged into default aswell.